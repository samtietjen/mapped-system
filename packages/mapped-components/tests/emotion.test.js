import React from "react";
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import { matchers } from 'jest-emotion';
import { render, cleanup } from '@testing-library/react';
import createUseMapper from '../src';

const useMapper = createUseMapper({
  breakpoints: [
    { label: null, minWidth: 0 }, 
    { label: "md", minWidth: '600px' },
    { label: "lg", minWidth: '1000px' }
  ],
  getter: ({ breakpoint, root, value }) =>
    [breakpoint, root, value]
      .filter(x => x && value !== false || x === 0)
      .join('-')
});

export const Box = useMapper({
  size: 'box-size'
});

Box.cssProps = {
  p: 'padding',
  m: 'margin',
  my: ['margin-top', 'margin-bottom']
}

Box.propTypes = {
  size: PropTypes.any,
  p: PropTypes.any,
  m: PropTypes.any,
  my: PropTypes.any
}

afterEach(cleanup);

expect.extend(matchers);

const createJson = Comp => renderer.create(Comp).toJSON();

describe('Mapped Components with Emotion', () => {
  test('Renders single properties.', () => {
    const json = createJson(<Box p="10px" m="100px" />);
    expect(json).toHaveStyleRule('padding', '10px');
    expect(json).toHaveStyleRule('margin', '100px');
    expect(json).toMatchSnapshot();
  });

  test('Renders responsive styles.', () => {
    const json = createJson(<Box m={["10px", "20px", "30px"]} />);
    expect(json).toHaveStyleRule('margin', '10px');
    expect(json).toHaveStyleRule('margin', '20px', { media: '(min-width: 600px)' });
    expect(json).toHaveStyleRule('margin', '30px', { media: '(min-width: 1000px)' });
    expect(json).toMatchSnapshot();  
  });

  test('Adds mapped class names.', () => {
    const a = createJson(<Box size="large" />);
    const b = createJson(<Box size={[1, 2, 3]} />);
    const classNames = a.props.className.split(' ');
    const responsiveClassNames = b.props.className.split(' ');
    expect(classNames.includes('box-size-large')).toBeTruthy();
    expect(responsiveClassNames.includes('box-size-1')).toBeTruthy();
    expect(responsiveClassNames.includes('md-box-size-2')).toBeTruthy();
    expect(responsiveClassNames.includes('lg-box-size-3')).toBeTruthy();
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
  });

  test('Includes element attributes.', () => {
    const json = createJson(<Box href="#" />);
    expect(json.props.href).toEqual('#');
    expect(json).toMatchSnapshot();
  });

  test('Forwards ref.', () => {
    const ref = React.createRef(null);
    const tree = render(<Box ref={ref} />);
    expect(ref.current.tagName).toBe('DIV');
  });
});