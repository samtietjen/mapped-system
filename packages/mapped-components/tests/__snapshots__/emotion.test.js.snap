// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapped Components with Emotion Adds mapped class names. 1`] = `
<div
  className="box-size-large css-7trrcd-MappedComponent e1oag0i80"
/>
`;

exports[`Mapped Components with Emotion Adds mapped class names. 2`] = `
<div
  className="box-size-1 md-box-size-2 lg-box-size-3 css-7trrcd-MappedComponent e1oag0i80"
/>
`;

exports[`Mapped Components with Emotion Includes element attributes. 1`] = `
<div
  className="css-7trrcd-MappedComponent e1oag0i80"
  href="#"
/>
`;

exports[`Mapped Components with Emotion Renders responsive styles. 1`] = `
<div
  className="css-bu69j4-MappedComponent e1oag0i80"
/>
`;

exports[`Mapped Components with Emotion Renders single properties. 1`] = `
<div
  className="css-1pqchow-MappedComponent e1oag0i80"
/>
`;
